scapy/contrib/pnio.py

```diff
@@ -41,37 +47,293 @@
     0xFF41:"PTCP-DelayResPDU-followup",
     0xFF42:"PTCP-DelayFuResPDU",
     0xFF43:"PTCP-DelayResPDU",
-    }
-for i in xrange(0x0100, 0x1000):
-    PNIO_FRAME_IDS[i] = "RT_CLASS_3"
-for i in xrange(0x8000, 0xC000):
-    PNIO_FRAME_IDS[i] = "RT_CLASS_1"
-for i in xrange(0xC000, 0xFC00):
-    PNIO_FRAME_IDS[i] = "RT_CLASS_UDP"
-for i in xrange(0xFF80, 0xFF90):
-    PNIO_FRAME_IDS[i] = "FragmentationFrameID"
+}
+
+
+def i2s_frameid(x):
+    if x in PNIO_FRAME_IDS:
+        return PNIO_FRAME_IDS[x]
+    elif 0x0100 <= x < 0x1000:
+        return "RT_CLASS_3 (%4x)" % x
+    elif 0x8000 <= x < 0xC000:
+        return "RT_CLASS_1 (%4x)" % x
+    elif 0xC000 <= x < 0xFC00:
+        return "RT_CLASS_UDP (%4x)" % x
+    elif 0xFF80 <= x < 0xFF90:
+        return "FragmentationFrameID (%4x)" % x
+    return x
+
+
+def s2i_frameid(x):
+    try:
+        idx = PNIO_FRAME_IDS.values().index(x)
+        return PNIO_FRAME_IDS.keys()[idx]
+    except ValueError:
+        pass
+    if x == "RT_CLASS_3":
+        return 0x0100
+    elif x == "RT_CLASS_1":
+        return 0x8000
+    elif x == "RT_CLASS_UDP":
+        return 0xC000
+    elif x == "FragmentationFrameID":
+        return 0xFF80
+    return x
 
 #################
 ## PROFINET IO ##
 #################
 
-class ProfinetIO(Packet):
+class ProfinetIO(scapy.packet.Packet):
     """Basic PROFINET IO dispatcher"""
-    fields_desc = [XShortEnumField("frameID", 0, PNIO_FRAME_IDS)]
-    overload_fields = {
-        Ether: {"type": 0x8892},
-        UDP: {"dport": 0x8892},
-        }
+    fields_desc = [
+        scapy.fields.XShortEnumField("frameID", 0, (i2s_frameid, s2i_frameid))
+    ]
 
     def guess_payload_class(self, payload):
         # For frameID in the RT_CLASS_* range, use the RTC packet as payload
-        if (self.frameID >= 0x0100 and self.frameID < 0x1000) or \
-                (self.frameID >= 0x8000 and self.frameID < 0xFC00):
-            from scapy.contrib.pnio_rtc import PNIORealTime
-            return PNIORealTime
+        if (0x0100 <= self.frameID < 0x1000) or (0x8000 <= self.frameID < 0xFC00):
+            return PNIORealTimeCyclicPDU
+        return super(ProfinetIO, self).guess_payload_class(payload)
+
+scapy.packet.bind_layers(scapy.layers.l2.Ether, ProfinetIO, type=0x8892)
```

Comment:
@gpotter2 said:
> you are not using the pythonish way of importing...
Indeed. I am cursed with the knownledge of many programming languages. That means that I often does not write code following a specific language philosophy. Instead, I spend much time thinking about how to write robust (or sometimes secure) code by borrowing useful idioms from other languages.

> Moreover, is you use a 'static' way of importing stuff using scapy.something..., what is imported ? What isn't ?
> Using a static approach scapy.arch.somefunction, are you sure that every thing will be loaded correctly ?

Doing `import <library>` executes all the statements in the code path of the library (including class and function definitions). If there are conditional statements at the global scope, only some code pathes are explored. The difference with `from <library> import <symbol>` is that you don't use and import one of the module symbols into the importing module. That is why the `import <library>` approach "always" works, while `from <library> import <symbol>` does not, in some instances.
Also, while doing imports inside a function might work, it actually intersperses the imports, which also hinders readability. That is why in my projects, I strongly object with this practice.

> As Python can import anything at anytime it wants to, unlike other languages like C, Java...
This statement is inaccurate. Doing an import in the middle of a function is similar to dynamic loading, which is possible in most languages.

@p-l said:
> My 2 cts: it's better to write from scapy.layers import bind_layers, [...] at the beginning of the code.

I disagree with "it's better", but I will update the PR to comply with Scapy requirements. Thank you.